exporter/fileexporter/grouping_file_exporter.go

```diff
@@ -0,0 +1,277 @@
+// Copyright The OpenTelemetry Authors
+// SPDX-License-Identifier: Apache-2.0
+
+package fileexporter // import "github.com/open-telemetry/opentelemetry-collector-contrib/exporter/fileexporter"
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	"os"
+	"path"
+	"strings"
+	"sync"
+
+	"github.com/hashicorp/golang-lru/v2/simplelru"
+	"go.opentelemetry.io/collector/component"
+	"go.opentelemetry.io/collector/consumer/consumererror"
+	"go.opentelemetry.io/collector/pdata/pcommon"
+	"go.opentelemetry.io/collector/pdata/plog"
+	"go.opentelemetry.io/collector/pdata/pmetric"
+	"go.opentelemetry.io/collector/pdata/ptrace"
+	"go.uber.org/zap"
+)
+
+type groupingFileExporter struct {
+	conf          *Config
+	logger        *zap.Logger
+	marshaller    *marshaller
+	pathPrefix    string
+	pathSuffix    string
+	attribute     string
+	maxOpenFiles  int
+	newFileWriter func(path string) (*fileWriter, error)
+
+	mutex   sync.Mutex
+	writers *simplelru.LRU[string, *fileWriter]
+}
+
+func (e *groupingFileExporter) consumeTraces(ctx context.Context, td ptrace.Traces) error {
+	if td.ResourceSpans().Len() == 0 {
+		return nil
+	}
+
+	groups := make(map[string][]ptrace.ResourceSpans)
+
+	for i := 0; i < td.ResourceSpans().Len(); i++ {
+		rSpans := td.ResourceSpans().At(i)
+		group(e, groups, rSpans.Resource(), rSpans)
+	}
+
+	var errs error
+	for pathSegment, rSpansSlice := range groups {
+		traces := ptrace.NewTraces()
+		for _, rSpans := range rSpansSlice {
+			rSpans.CopyTo(traces.ResourceSpans().AppendEmpty())
+		}
+
+		buf, err := e.marshaller.marshalTraces(traces)
+		if err != nil {
+			errs = errors.Join(errs, err)
+			continue
+		}
+
+		err = e.write(ctx, pathSegment, buf)
+		if err != nil {
+			errs = errors.Join(errs, err)
+		}
+	}
+
+	if errs != nil {
+		return consumererror.NewPermanent(errs)
+	}
+
+	return nil
+}
+
+func (e *groupingFileExporter) consumeMetrics(ctx context.Context, md pmetric.Metrics) error {
+	if md.ResourceMetrics().Len() == 0 {
+		return nil
+	}
+
+	groups := make(map[string][]pmetric.ResourceMetrics)
+
+	for i := 0; i < md.ResourceMetrics().Len(); i++ {
+		rMetrics := md.ResourceMetrics().At(i)
+		group(e, groups, rMetrics.Resource(), rMetrics)
+	}
+
+	var errs error
+	for pathSegment, rMetricsSlice := range groups {
+		metrics := pmetric.NewMetrics()
+		for _, rMetrics := range rMetricsSlice {
+			rMetrics.CopyTo(metrics.ResourceMetrics().AppendEmpty())
+		}
+
+		buf, err := e.marshaller.marshalMetrics(metrics)
+		if err != nil {
+			errs = errors.Join(errs, err)
+			continue
+		}
+
+		err = e.write(ctx, pathSegment, buf)
+		if err != nil {
+			errs = errors.Join(errs, err)
```

Comment:
Permanent error here as well.

Reply:
oops, long day

Reply:
The correct place for wrapping it with NewPermanent is a couple lines below, see:

https://github.com/open-telemetry/opentelemetry-collector-contrib/pull/31396/files#diff-2c93a6fc48837207b37ca4f2224563b7b131a435089dbdbe5acd165318758c42R109

Reply:
Ah, thanks.